include/posix/sys/types.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __POSIX_TYPES_H__
+#define __POSIX_TYPES_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include_next <sys/types.h>
+
+#ifdef CONFIG_PTHREAD_IPC
+#include <kernel.h>
+
+/* Thread attributes */
+typedef struct pthread_attr_t {
+	int priority;
+	void *stack;
+	size_t stacksize;
+	u32_t flags;
+	u32_t delayedstart;
+	u32_t schedpolicy;
+	s32_t detachstate;
+	u32_t initialized;
+} pthread_attr_t;
+
+typedef void *pthread_t;
+
+/* Mutex */
+typedef struct pthread_mutex {
+	struct k_sem *sem;
+} pthread_mutex_t;
```

Comment:
why is this not  "typedef struct k_sem pthread_mutex_t"

Reply:
@andyross  is the right person to comment on this, it was implemented by him. As per POSIX requirement all types should be inside types.h file. So I mode it from pthread.h --> types.h file

Reply:
It's just a typesafety thing.  If you typedef it then it becomes legal to pass a semaphore to pthread_mutex_unlock(), which doesn't work in general.  You can implement a mutex in terms of a semaphore, but not the reverse.

Reply:
Oh wait, is the question why it's a pointer and not a simple struct inclusion?  It has to do with our dlist declaration semantics: a zero-filled k_sem object is not a legal object.  You have to declare it with e.g. K_SEM_DEFINE.  So PTHREAD_MUTEX_DEFINE either needs to duplicate that code or use the existing macro, which requires that it store a pointer.

I had a patch to allow dlist to interpret NULL fields as "empty" (not zero cost, sadly) which would fix that.  Not sure if it's worth revisiting, or alternatively just fix pthread_mutex_init() to initialize the semaphore/dlist fields directly.  We should file a bug.

Reply:
Uh... ignore what I wrote.  We need to fix this now.

As written (yeah, by me), pthread_mutex_init() is going to call k_sem_init() on a garbage pointer on any mutex object that was not created with PTHREAD_MUTEX_DEFINE (which is itself a zephyr-specific API).   That's a pretty bad bug.  We should make the field an included struct, not a pointer, and change PTHREAD_MUTEX_DEFINE so that it initializes the dlist fields the same way K_SEM_DEFINE does.